runtime-sdk-macros/src/method_handler_derive.rs

```diff
@@ -0,0 +1,456 @@
+            (None, None, None, Some(rpc_name)) => (rpc_name, HandlerKind::Prefetch),
+            (_, _, _, _) => {
+                emit_compile_error(
+                    "#[handler] must specify exactly one of `call`, `query`, `message_result`, `prefetch` attributes",
+                );
+                return None;
+            }
+        };
+        Some(HandlerInfo {
+            kind: handler_kind,
+            rpc_name,
+            ident: ident.clone(),
+        })
+    }
+
+    let (handlers, nonhandlers): (Vec<ParsedImplItem>, Vec<ParsedImplItem>) = impl_block
+        .items
+        .into_iter()
+        .map(|item| ParsedImplItem {
+            handler: maybe_parse_handler(&item),
+            item,
+        })
+        .partition(|p| p.handler.is_some());
+
+    let handler_items = handlers
+        .iter()
+        .map(|ParsedImplItem { item, .. }| item)
+        .collect::<Vec<_>>();
+    let nonhandler_items = nonhandlers
+        .into_iter()
+        .map(|ParsedImplItem { item, .. }| item)
+        .collect::<Vec<_>>();
+
+    /// Generates parallel vectors of rpc names and handler idents for all handlers of kind `kind`.
+    fn filter_by_kind(
+        handlers: &[ParsedImplItem],
+        kind: HandlerKind,
+    ) -> (Vec<String>, Vec<syn::Ident>) {
+        handlers
+            .iter()
+            .filter_map(|h| h.handler.as_ref())
+            .filter(|h| h.kind == kind)
+            .map(|h| (h.rpc_name.clone(), h.ident.clone()))
+            .unzip()
+    }
+
+    let prefetch_impl = {
+        let (handler_names, handler_idents) = filter_by_kind(&handlers, HandlerKind::Prefetch);
+
+        // Find call handlers; for every call handler without a corresponding prefetch handler, we'll
+        // generate a dummy prefetch handler.
+        let (call_handler_names, _) = filter_by_kind(&handlers, HandlerKind::Call);
+        let handler_names_without_impl: Vec<&String> = call_handler_names
+            .iter()
+            .filter(|n| !handler_names.contains(n))
+            .collect();
+
+        if handler_names.is_empty() {
+            quote! {}
+        } else {
+            quote! {
+                fn prefetch(
+                    prefixes: &mut BTreeSet<Prefix>,
+                    method: &str,
+                    body: cbor::Value,
+                    auth_info: &AuthInfo,
+                ) -> module::DispatchResult<cbor::Value, Result<(), error::RuntimeError>> {
+                    let mut add_prefix = |p| {prefixes.insert(p);};
+                    match method {
+                        // "Real", user-defined prefetch handlers.
+                        #(
+                          #handler_names => module::DispatchResult::Handled(
+                            Self::#handler_idents(&mut add_prefix, body, auth_info)
+                          ),
+                        )*
+                        // No-op prefetch handlers.
```

Comment:
I was following the pattern from https://github.com/oasisprotocol/oasis-sdk/blob/main/runtime-sdk/src/modules/consensus_accounts/mod.rs#L374, where a no-op prefetch is explicit. It's also explicitly enforced by unit tests in that module: https://github.com/oasisprotocol/oasis-sdk/blob/main/runtime-sdk/src/modules/consensus_accounts/test.rs#L895

But I do see what you're talking about:
https://github.com/oasisprotocol/oasis-sdk/blob/main/runtime-sdk/src/dispatcher.rs#L300

Should I remove the autogenerated no-op handlers and update the test in consensus_accounts to be OK with unhandled prefetch?

Reply:
The difference between explicitly (no-op) handled and `Unhandled` is that in case of the explicitly handled, the `Tuple` implementation will return early:

https://github.com/oasisprotocol/oasis-sdk/blob/ba49e50b216f471257436e560f551f061003bfce/runtime-sdk/src/module.rs#L200-L206

This should be irrelevant in practice.

---

The only reason that Unhandled prefetch is allowed is mostly due to this being a bit of an "experimental" feature at the moment and is not widely used (also disabled by default) so only a couple of modules were provided with the implementation.

I think there are two equally valid steps we can take here:
- keep the autogenerated no-op handlers and don't allow unhandled prefetch
- allow unhandled prefetch, remove the autogenerated no-op handlers

Unless I missed/forgot any important other details, I think I would prefer the first option (keep the autogenerated handlers).

Reply:
I see, ok agree that the first option (keeping autogenerated no-op handlers) is better. And we can forbid unhandled prefetch separately.

Reply:
Thank you for the explanation, Peter. I'm keeping the autogenerated no-op handlers then.